### DISCUSSION TITLE
Segmentation Fault when "killing" a gif image

### DISCUSSION THREAD

**Discussion by maleblond** [libvips/libvips#2707] [Help wanted]
Hi,

We're setting a timeout to limit how long vips command takes (i.e. if it's taking more than x seconds, just kill the image). However, when that happens for a GIF, it causes a segmentation fault.

We compiled libvips from source from [this commit](https://github.com/libvips/libvips/commit/ac35124f590d7cfff927a161da723fa9e9c0d351), with cgif 0.2.1 (also built from source).

I wrote a little C program to reproduce the issue: https://gist.github.com/maleblond/f7f2851f4f43a8c19b9ec5c94eb3c8e8 . Forgive my poor C coding skills 😅  I've put a 1 second timeout to reproduce the issue(https://gist.github.com/maleblond/f7f2851f4f43a8c19b9ec5c94eb3c8e8#file-giftimeout-c-L67).

The program receives the path to a GIF image as the first argument (i.e. `giftimeout ./sample.gif`). Here is a beautiful dolphin GIF I was using for testing: https://user-images.githubusercontent.com/11162235/156651372-3e6cc842-883c-444c-b452-799547964091.gif

I used `gdb` to debug the segmentation fault, and here's the backtrace I was able to get:

```
#0  0x000000400228c3c2 in  () at /lib/x86_64-linux-gnu/libc.so.6
#1  0x0000004001aaba28 in vips_target_write_unbuffered () at /opt/imagery4/lib/libvips.so.42
#2  0x0000004001aabf5f in vips_target_write () at /opt/imagery4/lib/libvips.so.42
#3  0x00000040033f9dc4 in cgif_raw_addframe (pGIF=<optimized out>, pConfig=pConfig@entry=0x4001805130) at ../src/cgif_raw.c:559
#4  0x00000040033f87dd in flushFrame (pGIF=pGIF@entry=0x4008025640, pCur=<optimized out>, pBef=<optimized out>) at ../src/cgif.c:338
#5  0x00000040033f9199 in cgif_close (pGIF=0x4008025640) at ../src/cgif.c:459
#6  0x0000004001a456c9 in vips_foreign_save_cgif_dispose () at /opt/imagery4/lib/libvips.so.42
#7  0x0000004002036c93 in g_object_unref () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#8  0x0000004001ac7a8a in vips_call_split () at /opt/imagery4/lib/libvips.so.42
#9  0x0000004001a473b1 in vips_gifsave_buffer () at /opt/imagery4/lib/libvips.so.42
#10 0x0000004000001393 in main (argc=<optimized out>, argv=<optimized out>) at giftimeout.c:110
```

This seems to point at `cgif` trying to write to something that was already freed, but we aren't quite sure. LMK if you need any more information. Any help would be greatly appreciated!

**Comment by jcupitt** [comment]
Thanks for the bug report! Fixed here: https://github.com/libvips/libvips/issues/2708
